design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+  installed by the package.
+- `spec.compositions`: list of `Composition`s installed by the package.
+- `spec.state`: the desired state, which can be one of `Pending`, `Active`,
+  `Stopping` or `Inactive`.
+
+The following fields will be removed:
+
+- `spec.packageType`: currently is one of `Provider`, `Stack`, `Application`, or
+  `Addon`.
+- `spec.permissionType`: currently is one of `Cluster` or `Namespace`.
+
+In addition, `PackageRevision` will be a cluster-scoped type, rather than the
+current namespace-scoped `Package`. The current schema for `Package` can be
+viewed
+[here](https://doc.crds.dev/github.com/crossplane/crossplane/cluster/charts/crossplane-types/crds/packages.crossplane.io_packages.yaml@v0.12.0).
+
+In order to continue to support the current functionality of
+`ClusterPackageInstall`, `PackageInstall`, `StackDefinition` the underlying
+`Package` type will be renamed to `LegacyPackage`. The existing `Package` type
+will be refactored to look like the current `ClusterPackageInstall` schema with
+thw following modifications.
+
+This field will be added:
+
+- `status.currentHash`: a string that is the `sha256` hash of the image in the
+  current `PackageRevision`.
+- `status.currentImage`: a string that is the name of the current
+  `PackageRevision` image.
+
+The following fields will be removed:
+
+- `status.installJob`: a reference to the install `Job` for the
+  `ClusterPackageInstall`. 
+- `status.packageRecord`: a reference to the `Package` that represents the
+  installation unit for the `ClusterPackageInstall`. The `PackageRevision` is
+  now discoverable by name at the cluster scope.
+
+
+Before outlining the functionality of the new `Package` and `PackageRevision`
+controllers, it is useful to establish the responsibilities of each.
+
+The `Package` controller is responsible for:
+
+- Managing the lifecycle of its `PackageRevision`s
+- Managing the lifecycle of any objects installed by its `PackageRevision`s
+
+The `PackageRevision` controller is responsible for:
+
+- Unpacking a specific package image
+- Managing the lifecycle of any controller is installed by a package image
+- Managing the ownership of objects included in the package
+
+Note that the `Package` controller is responsible for resources that may persist
+between `PackageRevision`s (CRDs, `InfrastructureDefinition`s,
+`InfrastructurePublication`s, `Composition`s), while the `PackageRevision`
+controller is only responsible for resources that are specific to that revision
+(`Deployment`, `Job`) and _ownership_ of the resources that persist.
+
+#### New Package Controller
+
+The new _`Package` controller_ will observe Create / Update / Delete operations
+on a `Package` and create `PackageRevision`s in response. `PackageRevision`s
+should be given the name of the `sha256` hash of the image provided in the
+`Package` `spec.package` field. Each change to a `Package` that results in a new
```

Comment:
Where does the sha256 hash come from? The user only supplies the image name/tag and registry source, is the sha256 hash something that the Package controller asks from the source registry? or does it compute it?

Reply:
@jbw976 mentioned a bit below, the `Job` reports the hash in its status: https://github.com/crossplane/crossplane/pull/1616#discussion_r448350362